File: src/main/java/beholder_daemon/utils/GzipUtils.java
package beholder_daemon.utils;

import java.io.*;
import java.util.zip.*;
import org.brotli.dec.BrotliInputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;


public class GzipUtils {
    
    // 压缩
    public static byte[] compress(String data) throws IOException {
        if (data == null || data.isEmpty()) return null;
        
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        try (GZIPOutputStream gzip = new GZIPOutputStream(bos)) {
            gzip.write(data.getBytes("UTF-8"));
        }
        return bos.toByteArray();
    }
    
    // 解压
    public static String decompress(byte[] compressed) throws IOException {
        if (compressed == null) return null;
        
        ByteArrayInputStream bis = new ByteArrayInputStream(compressed);
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        try (GZIPInputStream gzip = new GZIPInputStream(bis)) {
            byte[] buffer = new byte[1024];
            int len;
            while ((len = gzip.read(buffer)) != -1) {
                bos.write(buffer, 0, len);
            }
        }
        return bos.toString("UTF-8");
    }

    public static byte[] decompressBrotli(byte[] compressed) throws IOException {
        try (ByteArrayInputStream bis = new ByteArrayInputStream(compressed);
            BrotliInputStream brotli = new BrotliInputStream(bis);
            ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
            
            byte[] buffer = new byte[1024];
            int len;
            while ((len = brotli.read(buffer)) != -1) {
                bos.write(buffer, 0, len);
            }
            return bos.toByteArray();
        }
    }
}

File: internal/utils/gzip_utils.go
package utils

import (
	"bytes"
	"compress/gzip"
	"io"
)

// GzipUtils 压缩工具类
type GzipUtils struct{}

// NewGzipUtils 创建 GzipUtils 实例
func NewGzipUtils() *GzipUtils {
	return &GzipUtils{}
}

// Compress 压缩字符串为字节数组
func (g *GzipUtils) Compress(data string) ([]byte, error) {
	if data == "" {
		return nil, nil
	}

	var buf bytes.Buffer
	gzipWriter := gzip.NewWriter(&buf)

	_, err := gzipWriter.Write([]byte(data))
	if err != nil {
		return nil, err
	}

	if err := gzipWriter.Flush(); err != nil {
		return nil, err
	}

	if err := gzipWriter.Close(); err != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}

// Decompress 解压缩字节数组为字符串
func (g *GzipUtils) Decompress(compressed []byte) (string, error) {
	if compressed == nil || len(compressed) == 0 {
		return "", nil
	}

	buf := bytes.NewReader(compressed)
	gzipReader, err := gzip.NewReader(buf)
	if err != nil {
		return "", err
	}
	defer gzipReader.Close()

	var out bytes.Buffer
	_, err = io.Copy(&out, gzipReader)
	if err != nil {
		return "", err
	}

	return out.String(), nil
}

// DecompressBrotli 解压缩 Brotli 压缩的数据
func (g *GzipUtils) DecompressBrotli(compressed []byte) ([]byte, error) {
	if compressed == nil || len(compressed) == 0 {
		return nil, nil
	}

	// 注意：这里需要使用 brotli 库，暂时使用 gzip 替代
	// 实际使用时需要引入 github.com/andybalholm/brotli
	buf := bytes.NewReader(compressed)
	gzipReader, err := gzip.NewReader(buf)
	if err != nil {
		return nil, err
	}
	defer gzipReader.Close()

	var out bytes.Buffer
	_, err = io.Copy(&out, gzipReader)
	if err != nil {
		return nil, err
	}

	return out.Bytes(), nil
}
